store/store.go

```diff
@@ -1109,21 +1103,31 @@ func (s *Store) Download(name string, downloadInfo *snap.DownloadInfo, pbar prog
 		url = downloadInfo.DownloadURL
 	}
 
-	if err := download(name, url, user, s, w, pbar); err != nil {
-		return "", err
+	sha3_384, err := download(name, url, user, s, w, pbar)
+	if err != nil {
+		return err
+	}
+
+	if downloadInfo.Sha3_384 != "" && sha3_384 != downloadInfo.Sha3_384 {
+		// FIXME: find a better error message
+		return fmt.Errorf("hashsum mismatch for %s: got %s but expected %s", w.Name(), sha3_384, downloadInfo.Sha3_384)
 	}
 
-	return w.Name(), w.Sync()
+	if err := os.Rename(w.Name(), targetFn); err != nil {
+		return err
+	}
+
+	return w.Sync()
 }
 
 // 3 pₙ₊₁ ≥ 5 pₙ; last entry should be 0 -- the sleep is done at the end of the loop
 var downloadBackoffs = []int{113, 191, 331, 557, 929, 0}
 
 // download writes an http.Request showing a progress.Meter
-var download = func(name, downloadURL string, user *auth.UserState, s *Store, w io.Writer, pbar progress.Meter) error {
+var download = func(name, downloadURL string, user *auth.UserState, s *Store, w io.Writer, pbar progress.Meter) (sha3_384 string, err error) {
```

Comment:
Same as the comment in the other PR, can we please revert the logic so we return the error above from within download, and pass the sha3-384 into `download` so it can do the comparison internally and report the error consistently for all cases.